pkg/chunk/cache/redis_cache.go

```diff
@@ -0,0 +1,244 @@
+package cache
+
+import (
+	"context"
+	"errors"
+	"flag"
+	"time"
+
+	"github.com/cortexproject/cortex/pkg/util"
+	"github.com/go-kit/kit/log/level"
+	"github.com/gomodule/redigo/redis"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	instr "github.com/weaveworks/common/instrument"
+)
+
+// RedisClient interface exists for mocking redisClient.
+type RedisClient interface {
+	Connection() redis.Conn
+	Close() error
+}
+
+// RedisCache type caches chunks in redis
+type RedisCache struct {
+	name            string
+	expiration      int
+	timeout         time.Duration
+	client          RedisClient
+	requestDuration observableVecCollector
+}
+
+// RedisConfig defines how a RedisCache should be constructed.
+type RedisConfig struct {
+	Endpoint       string        `yaml:"endpoint,omitempty"`
+	Timeout        time.Duration `yaml:"timeout,omitempty"`
+	Expiration     time.Duration `yaml:"expiration,omitempty"`
+	MaxIdleConns   int           `yaml:"max_idle_conns,omitempty"`
+	MaxActiveConns int           `yaml:"max_active_conns,omitempty"`
+}
+
+type redisClient struct {
+	pool *redis.Pool
+}
+
+var (
+	redisRequestDuration = promauto.NewHistogramVec(prometheus.HistogramOpts{
+		Namespace: "cortex",
+		Name:      "redis_request_duration_seconds",
+		Help:      "Total time spent in seconds doing redis requests.",
+		// Redis requests are very quick: smallest bucket is 16us, biggest is 1s
+		Buckets: prometheus.ExponentialBuckets(0.000016, 4, 8),
+	}, []string{"method", "status_code", "name"})
+
+	errRedisQueryTimeout = errors.New("redis query timeout")
+)
+
+// RegisterFlagsWithPrefix adds the flags required to config this to the given FlagSet
+func (cfg *RedisConfig) RegisterFlagsWithPrefix(prefix, description string, f *flag.FlagSet) {
+	f.StringVar(&cfg.Endpoint, prefix+"redis.endpoint", "", description+"Redis service endpoint to use when caching chunks. If empty, no redis will be used.")
+	f.DurationVar(&cfg.Timeout, prefix+"redis.timeout", 100*time.Millisecond, description+"Maximum time to wait before giving up on redis requests.")
```

Comment:
Is there some way we could share these flags with memcached?
E.g. have a memcached name and a neutral name both pointing to the same variable so we can deprecate and eventually remove the cache-specific names?

Reply:
I agree that it doesn't make sense to have both `redis` and `memcached`. I should fix that.
For example, we can have `--cache.store.type=(memcached|redis)` parameter to select the cache type.
But could you elaborate about cache-agnostic parameters? Some parameters are cache-specific, for example `batch_size` in `memcached`. How do you envision the command line would look like?
Do you mean something like that:
```
--cache.store.type=memcached --cache.store.batch_size=0 --cache.store.parallelism=10

--cache.store.type=redis --cache.store.host=redis.default:6379
```